### ISSUE TITLE
Footnotes aren't moved when the last line of the page is moved

### ISSUE THREAD

**Issue by hpvd** [gucci-on-fleek/lua-widow-control#26]
### Issue

was just playing around with lua-widow-control and some borrowed code...

There is a question:
what happens if lua-widow-control moves with its ongoing work a reference of a footnote to the next page?
There seems to be a chance leading to a footnote shown/starting on the wrong page (the page before).
edit: or does one start an endless loop of resetting with this?

![2022-03-13_21h09_26](https://user-images.githubusercontent.com/5681880/158077742-0706812a-8dfc-4c44-aae7-20f8ecd85a71.png)

### Expected Behavior

would expect that the footnote start at the page where its reference is used

### Reproduction

```latex
\documentclass{article}

\setlength{\parindent}{5.0mm}
\setlength{\parskip}{0mm}
\usepackage[paperwidth=140mm,paperheight=213mm]{geometry}

\usepackage{lipsum}

\usepackage{lua-widow-control}

\begin{document}

\lipsum*[1-2]
\footnote{\lipsum*[1-2]} This is some text finishing this paragraph that reaches to the next line.

\lipsum*[2-7]
\end{document}
```

### Format

LaTeX

### Lua-widow-control Version

v2.0

**Comment by hpvd** [comment]
hmm just noticed that the page or the footnote is shortend:
![2022-03-13_21h09_261](https://user-images.githubusercontent.com/5681880/158078422-4da3913e-a702-4ff4-8d44-9c22f6bf8255.png)

**Comment by hpvd** [comment]
I'm sorry, I just can't show the behavior without lua-widow-control, there is a font problem on my machine when trying to open it with a pdf reader having a  twoside view...

Can you reproduce the  shown above?

**Comment by hpvd** [comment]
this may lead to a general question regarding interaction of lua-widow-control with references to any kind of object (footnote and different floats) which were moved  to a different page
and
the position of the target of these references (footnotes, images, tables etc.)

Should these work fine (in theory) or is it a thing one have to think of?

**Comment by hpvd** [comment]
for comparison:
same document without lua-widow-control:

![2022-03-13_22h40_54](https://user-images.githubusercontent.com/5681880/158080459-dd48f3d5-69b9-4fa4-80d2-20c69531c2d8.png)

**Comment by hpvd** [comment]
and with
\usepackage{lua-widow-control}
\usepackage{lua-visual-debug}

to help to understand what is going on at the bottom edge:
![2022-03-13_22h48_43](https://user-images.githubusercontent.com/5681880/158080696-894e58cb-5234-4d8d-bc86-19a2d9e27533.png)

**Comment by hpvd** [comment]
and 
```
% \usepackage{lua-widow-control}
\usepackage{lua-visual-debug}
```
![2022-03-13_22h54_36](https://user-images.githubusercontent.com/5681880/158080864-cb6141cf-8bb5-42f5-906a-bd6717cc5b8b.png)

**Comment by hpvd** [comment]
to fix any warning of the given example code one should add
`\usepackage[base]{babel}`

but it doesn't change the result...

**Comment by gucci-on-fleek** [comment]
Thanks for the detailed report!

This is a fairly tricky issue to fix. I _should_ be able to get this fixed in a future update, but I probably won't have time until April/May. I'm tracking the fix in #27.

As a work around, you can look through the log file for lines like:
```log
Module lua-widow-control Info: Widow/orphan successfully removed at paragraph <#> on page <#>.
```
You can then manually check those pages to make sure that the footnotes didn't get separated. It's not much of a solution though. Hopefully I'll be able to get this fixed.

**Comment by hpvd** [comment]
Sorry for stumbling across such a big issue.

Don't worry about needing time for other/more important things.

just an idea for an intermediate fix to start with (in May)
maybe one could disable the work of lua-widow-control in a all cases where any type of reference occurs/would be moved to an other page?

This could be mentioned in Doc named as ToDo, an you got peace of mind and some month of time.
Maybe in the meantime someone has a good initial hint :-)

**Comment by hpvd** [comment]
maybe there is a different behavior if there is only a short footnote which does not spread to two pages.
will setup a minimal example for this within the next day

**Comment by hpvd** [comment]
the point "knowing" where a reference is placed, is also import for packages like varioref
which can do things like giving different texts for a reference, depending on if the target (float: image, table...) is on the same or next page.

> The package introduces commands \vref and \vpageref. The \vref command refers to labels in the same way as does \ref, but if the thing referred to is on a different page, the result will be decorated with “on the facing page” or “on page <n>”, and so on, as appropriate. The \vpageref command similarly decorates a page number; there are also commands \vrefrange and \vpagerefrange for ranges of labels. 

https://ctan.org/pkg/varioref

**Comment by hpvd** [comment]
packages like fewerfloatpages may also move floats to different places
maybe one should check if this happens at compatible point of compilation
https://ctan.org/pkg/fewerfloatpages

**Comment by gucci-on-fleek** [comment]
@hpvd 
> Sorry for stumbling across such a big issue.

Well, it's better that I know about the issue so that I can fix it. The issue exists regardless if anyone reports it.

> maybe one could disable the work of lua-widow-control in a all cases where any type of reference occurs/would be moved to an other page?

Detecting that the last line contains a footnote is already half the work, so that wouldn't speed things up too much. Once exams are over then I should have some time to work on this.

> This could be mentioned in Doc named as ToDo

It's on [the list](https://github.com/users/gucci-on-fleek/projects/1/views/5). I'll add it to the "Known Issues" section of the manual at some point.

> maybe there is a different behavior if there is only a short footnote which does not spread to two pages.

I don't think that it depends on the length of the footnote (although I could be wrong). The problem is that there isn't a LuaTeX callback that replaces page breaking, so lua-widow-control has to step in _after_ the page has already been broken. But at this point, the footnote has already been placed. The "footnote mark" is inline with the paragraph so lua-widow-control moves that. It doesn't account for the "footnote text" at all (yet—that's what this bug is).

The TeXbook pp. 122–125 and [TeX by Topic Chapter 29](https://texdoc.org/serve/texbytopic/0#chapter.29) are decent references here. Inserts/output routines are a really complicated topic that even I don't really understand.

> the point "knowing" where a reference is placed, is also import for packages like `varioref`

I think that `varioref` writes an object's location using a non-`\immediate` write into the `.aux` file (but I haven't actually checked). This shouldn't be triggered until the page is actually shipped out, which is well after lua-widow-control does anything. If you find any problems here, then please report it, but I think that it should be fine.

> packages like fewerfloatpages may also move floats to different places

Lua-widow-control shouldn't interfere with floats. A standard figure/table is just on the page, not embedded in a paragraph. Lua-widow-control (mostly) only hooks into paragraphs, so there shouldn't (hopefully) be an issue here.

**Comment by hpvd** [comment]
many thanks for your detailed answer!

Just to clarify, the point with floats is, that these shouldn't occur on a page where they haven't already been mentioned (referenced).

So if be before lua-widow-control comes into the game, the reference is in the last line of a page and the figure is right below (or above?) this,
by the work of lua-widow-control  the reference may move to the next page, but the float stays on the original page.
With this the float is set on a page before it was mentioned (referenced)

So _if_ I get the desired behavior for floats right, it looks like basically the same problem as with the footnotes, doesn't it?


GOOD LUCK with your examens!

**Comment by hpvd** [comment]
not sure, maybe this contains something useful:
https://www.latex-project.org/help/documentation/#documentation-of-the-new-hook-management

**Comment by gucci-on-fleek** [comment]
Ah, I see where you're coming from now.

The difference is is that the numbering for a footnote is always sequential, so if you have
```latex
\documentclass{article}
\begin{document}
    A\footnote{A} B\footnote{B} C\footnote{C} 
    … … …
\end{document}
```
then footnote A is always 1, footnote B is always 2 and so on, regardless of what comes after. 

On the other hand, if you have
```latex
\documentclass{article}
\begin{document}
    \begin{figure}[p]
        Hello!
        \caption{Figure}\label{figure}
    \end{figure}
    The figure is on page~\pageref{figure}.
    … … …
\end{document}
```
then you don't know which page the figure is going to be on since it depends on what comes after. If right after is `\clearpage`, then the figure will be on the next page; otherwise, it will wait until there are enough other figures to make a full float page, which could be in 10+ pages.

Now, TeX typesets 1 page at a time, without ever going back. With the footnotes, this is fine; their numbering doesn't depend on what comes next; however, the second example with the figure _does_ depend on what comes next. This isn't possible to solve in TeX, so LaTeX "cheats" and has the figure write its page number into the `.aux` file when it is actually typeset. Then, on a second run, the `\pageref` reads in from the `.aux` file.

With the footnotes, TeX assumes that the footnote number won't move after the page has been broken. This assumption is always correct, so long as you aren't using lua-widow-control. With floats, TeX makes no such assumption since it takes multiple passes for references to resolve properly.

---

The other difference here is that the footnote text insert is added when we are in the middle of a paragraph (`hmode`), while the figure content insert is added in between paragraphs (`vmode`). Lua-widow-control is only activated when it finds a widow or orphan, which means that the page ends halfway through a paragraph. It is impossible to insert a float in the middle of a paragraph, so lua-widow-control should never need to move the `\insert` location of a float.

The one other thing here is that unlike with footnotes, the reference to a float is unrelated to the float itself. So something like
```latex
\documentclass{article}
\begin{document}
    \begin{figure}
        Hello!
        \caption{Figure}\label{figure}
    \end{figure}
    \clearpage

    (100+ pages of text)

    As you can see in figure~\ref{figure}
    … … …
\end{document}
```
is completely valid (although quite questionable). 

So I think that this should be fine, but of course, let me know if you find any issues since I could be missing something.

---

> https://www.latex-project.org/help/documentation/#documentation-of-the-new-hook-management

Yep, I've got that in use in the LaTeX package:

https://github.com/gucci-on-fleek/lua-widow-control/blob/0f95988e1d8495aefd26fd7a7b3246840d5c5b73/source/lua-widow-control.sty#L117

It's not super useful though, since lua-widow-control works at a much lower layer than LaTeX. Lua-widow-control mainly uses LuaTeX callbacks. The official reference for those is the [LuaTeX manual](https://texdoc.org/serve/luatex/0#%231876). The [LuaTeX Wiki](http://wiki.luatex.org/index.php/Show_the_hyphenation_points) also has a pretty good example.

---

Oh, and if you want to see the page the same way that lua-widow-control sees it, you can try the [`nodetree` package](https://www.ctan.org/pkg/nodetree). It's pretty nifty, but also _really_ verbose. Only run it on a small document, otherwise you'll be overwhelmed.

**Comment by hpvd** [comment]
The misconception in my mind was, that I was pretty sure
- that with default settings latex would never put a float before it's position in source code and
- even further, not before its first reference (mad thought)
But both is not the case... obviously.

**Comment by hpvd** [comment]
in the first trials, varioref and lua-widow-control works fine together :-)

**Comment by gucci-on-fleek** [comment]
This is mostly fixed in [the new beta](https://github.com/gucci-on-fleek/lua-widow-control/releases/tag/release-5e4bdf417828294ad38c4804946cac471416560f). The only issue is that when a single footnote is split across multiple pages, the output is wrong. Otherwise, footnotes seem to work as expected now.

Once I fix the issue with the long footnotes, I'll push a release to CTAN.

**Comment by gucci-on-fleek** [comment]
This should be fixed in v2.1.2, which will be in TeX Live/MikTeX in about 2 days.

Let me know if you find any further issues. Thanks!
